packages/aragon-api/src/index.js

```diff
@@ -228,19 +308,74 @@ export class AppProxy {
     // (because of the `Promise.resolve`).
     const wrappedReducer = (state, event) =>
       from(
-        Promise.resolve(reducer(state, event))
+        // Ensure a promise is returned even if the reducer returns an array
+        Promise.resolve(reducer(state, event)).catch(e => console.error(
+          `Error from app reducer:`, e)
+        )
       )
 
-    const store$ = initialState.pipe(
-      switchMap((initialState) =>
-        merge(
-          this.events(),
-          ...events
-        ).pipe(
-          mergeScan(wrappedReducer, initialState, 1),
-          map((state) => this.cache('state', state))
+    const getCurrentEvents = (fromBlock) => merge(
+      this.events(fromBlock),
+      ...externals.map(({ contract }) => contract.events(fromBlock))
+    )
+
+    // If `cachedFromBlock` is null there's no cache, `pastEvents` will use the initialisationBlock
+    // External contracts can specify their own `initializationBlock` which will be used in case the cache is empty
+    const getPastEvents = (cachedFromBlock, toBlock) => merge(
```

Comment:
If we want to introduce a loading state for the fetching of past events, we probably want to use `concat` instead of `merge` and add a trigger/custom event before and after.

Reply:
We can use [`endWith()`](https://rxjs-dev.firebaseapp.com/api/operators/endWith) in this observable to emit this custom event.

`merge()` is nice so the reducer can start processing any of the contract events as they arrive.

---------

Alternatively (or in combination, both the reducer and an app's frontend may be interested in sync states), we may also want to use a separate API stream (e.g. `syncStatus()`) for this to not pollute the state cache. This would be really useful later on, when we fetch past blocks incrementally and could notify subscribers that we're past a certain stage, e.g.

```js
{
  status: 'SYNCING',
  details: {
    current: '<block number>',
    from: '<block number>',
    target: '<block number>',
  }
}

{
  status: 'SYNCED',
  details: {}
}
```

Reply:
Not directly related, but perhaps we should a tilde instead of caret for rxjs. I just hit a strange bug (https://jsbin.com/sasumoy/edit?html,js,console). It's already reported: https://github.com/ReactiveX/rxjs/issues/4731

Reply:
Hmm I think this type of bug is going to be pretty hard to avoid via a dependency lock (a patch version fixing a bug could also be buggy); the best we could do is add a `package-lock.json` (which _may_ be possible after upgrading web3.js), and of course, tests.

It seems like you might not have had the latest patch version either (rxjs@6.5.2 seems to have this fixed?), so as well, even with the `package-lock.json` we would've been screwed if it was locked to 6.5.1 🤷‍♂.

Reply:
Yep. `package-lock.json` is the way forward. You're right that we  would have been screwed if it were locked to 6.5.1 but at least we'd have been able to notice that the upgrade took place.